AArtur strategy

KingAArtur opened the pull request:
I have created my **super-duper** strategy. Approve `please`

KingAArtur committed:
AdvancedState and Map classes

KingAArtur committed:
Solver class for handling basic tasks

Solvers for collecting ScoreBonus and HealBonus and for shooting.

KingAArtur committed:
The actual strategy

AArturBaseStrategy just chooses from moves offered by its Solvers.
AArturSmartStrategy has the particular set of Solvers and coefficients.

KingAArtur committed:
Moved aartur strategy to strategies directory

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,97 @@
+        """Constructing a Map from State"""
+        self.size_x = state.size_x
+        self.size_y = state.size_y
+        self.map: list[list[MapCell]] = [[MapCell() for _ in range(state.size_x)] for _ in range(state.size_y)]
```
maybe MapCell()->None
type: list[list[MapCell | None]]

KingAArtur replied:
It is inconvinient, because I will have to replace all remaining `None` cells (which are unreachable) with a `MapCell()` manually

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,97 @@
+        self.size_y = state.size_y
+        self.map: list[list[MapCell]] = [[MapCell() for _ in range(state.size_x)] for _ in range(state.size_y)]
+
+        def fill_map(mapper: list[list[MapCell]], cells: list[tuple[int, int]], dist: int):
```
maybe you should make this a method

KingAArtur replied:
Done!

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,97 @@
+        self.visited: bool = visited
+
+
+class Map:
```
it looks like we can combine AdvancedState and Map in one class AdvancedState
because they are used only together and very similar

KingAArtur replied:
`Map` is used to determine how to move around the board for *one particular player*. I may want to create another `Map` for *another* player to look how he can move around the map. So I will have to duplicate the `AdvancedState` which is bad.

Mathematician2000 replied:
`Map` -> `ReachabilityGraph`? Or `BFSGraph`
`AdvancedState` -> `ExtendedState`?

polevgart replied:
ExtendedState must contain ReachabilityGraph, i think

KingAArtur replied:
Renamed to `ReachabilityGraph` and `ExtendedState`.
I don't think that State should contain Graph: they are used for different purposes and one may need Graphs for all players while other might don't need Graph at all.

polevgart commented on strategies/aartur/solver.py:
```diff
@@ -0,0 +1,151 @@
+
+        dist = mapper[y][x].dist
+        confidence = 1.0 - dist * 0.1
+        if self.bonus_type == HealBonus:
```
Сan we use redefinition after inheritance?
Like:
```
class CollectBonusSolver(BaseSolver):
  def solve(...):
    ...
    confidence = self.calculate_confidence(dist, player)
    ...

  @staticmethod
  def calculate_confidence(distance, player):
    raise NotImplementedError()

class CollectScoreBonusSolver(CollectBonusSolver):
  @staticmethod
  def calculate_confidence(distance, player):
    ...

class CollectHealBonusSolver(CollectBonusSolver):
  @staticmethod
  def calculate_confidence(distance, player):
    ...
```

KingAArtur replied:
Done!

polevgart commented on strategies/aartur/solver.py:
```diff
@@ -0,0 +1,151 @@
+        closest_shoot_positions: list[tuple[Player, tuple[int, int]]] = []
+        for enemy in state.other_players:
+            if not enemy.is_alive:
+                # ignoring dead enemies!
```
i think it's useless

KingAArtur replied:
No way i dont how know someone can live without this comment...
Done!

KingAArtur committed:
Defined get_cell and set_cell methods for AdvancedState and Map

Because using map[y][x] and state[y][x] is dangerous in comparison to state.get_cell(x=x, y=y)
Also moved recursive function 'fill_map' from __init__ to the _fill_map method

KingAArtur committed:
Defined calculate_confidence method in CollectBonusSolver

Getting rid of an ugly 'if' statement in solve method and allowing all inherited solvers to define how to calculate a confidence.

KingAArtur committed:
Renamed Map.goto method to get_direction_to

Also deleted this little innocent comment about shooting at dead enemies

KingAArtur committed:
Made AArturBaseStrategy.solvers instance attribute

Now all instances of AArturSmartStrategy can redefine their 'solvers' values without affecting other instances

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,104 @@
+        self.other_players: list[Player] = []
+        for y in range(self.size_y):
+            for x in range(self.size_x):
+                cell = self.cells[y][x]
```
get_cell

KingAArtur replied:
I will fix it with other minor fixes (don't wanna do the whole commit for that thing now)

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,104 @@
+        self._fill_map(state, [(state.player.x, state.player.y)], 1)
+
+    def _fill_map(self, state: AdvancedState, cells: list[tuple[int, int]], dist: int):
+        # Recursive function for filling map cells step-by-step
```
In fact, this is [bfs algorithm](https://en.wikipedia.org/wiki/Breadth-first_search)
it can be non recursive
:)

polevgart replied:
it can be non recursive :)

Mathematician2000 replied:
It **should** be non-recursive, use queue (deque?) instead

KingAArtur replied:
Yep. Now it is iterative with `deque`.

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,104 @@
+                    if (not isinstance(state.get_cell(x=new_x, y=new_y), (Player, Wall)) and
+                            not self.map[new_y][new_x].visited):
+                        next_cells.append((new_x, new_y))
+                        self.map[new_y][new_x] = MapCell(dist=dist, dx=dx, dy=dy, visited=True)
```
set_cell

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,104 @@
+                for dy in (-1, 0, 1):
+                    new_x, new_y = x + dx, y + dy
+                    if (not isinstance(state.get_cell(x=new_x, y=new_y), (Player, Wall)) and
+                            not self.map[new_y][new_x].visited):
```
get_cell

polevgart commented on strategies/aartur/preparation.py:
```diff
@@ -0,0 +1,104 @@
+        self.map: list[list[MapCell]] = [[MapCell() for _ in range(state.size_x)] for _ in range(state.size_y)]
+
+        # Initializing original cell with dist=0 and starting recursion process
+        self.map[state.player.y][state.player.x] = MapCell(dist=0, visited=True)
```
set_cell

polevgart commented on strategies/aartur/solver.py:
```diff
@@ -0,0 +1,165 @@
+        for enemy, (x, y) in closest_shoot_positions:
+            if mapper.get_cell(x=x, y=y).dist > 0:
+                # if distance > 0 then we cannot shoot directly, we have to walk
+                move = mapper.get_direction_to(x, y)
```
maybe mapper rename to graph? idk

Mathematician2000 replied:
According to new class name

KingAArtur replied:
Sure, renamed everywhere

polevgart commented on strategies/aartur/straregy.py:
```diff
@@ -0,0 +1,50 @@
+    solvers: dict[BaseSolver, float] = attr.ib(factory=dict, init=False)
+
+    def get_next_move(self, state: State) -> BaseMove:
+        moves: list[tuple[BaseMove, float]] = []  # if only the DirectMove and Shoot were hashable...
```
for what need hash of DirectMove and Shoot?

KingAArtur replied:
For more convinient usage such as using in `dict[BaseMove, something]` and checking the equality of moves.
Removed this comment

polevgart commented on strategies/aartur/straregy.py:
```diff
@@ -0,0 +1,50 @@
+
+        for solver, priority in self.solvers.items():
+            move, confidence = solver.solve(state, mapper)
+            moves.append((move, priority * confidence))
```
we can save only max value as result

Mathematician2000 replied:
Don't need a whole list to store the moves, neither you need tuples and hashable moves at the moment.
Consider rewriting it in a streaming fashion

KingAArtur replied:
I removed constructing the list from choosing the best move, but maybe i did it not in the perfect way. Check pls

KingAArtur committed:
Renamed AdvancedState to ExtendedState and Map to ReachabilityGraph

KingAArtur committed:
Made ReachabilityGraph constructing iterative (BFS with deque)

Also used get_cell and set_cell methods while constructing instead of direct list[y][x]

KingAArtur committed:
Choosing the best move in get_next_move without list creation

KingAArtur committed:
Using get_cell
